### PR TITLE
[FLINK-17334] HIVE UDF BUGFIX

### PR DESCRIPTION
**Pull request by Ruan-Xin**
For more information, please check
https://issues.apache.org/jira/browse/FLINK-17334#

### CODE REVIEW

**Comment by JingsongLi** [comment]
Can you add some tests for this?

**Comment by Ruan-Xin** [comment]
> Can you add some tests for this?

Test case:
https://github.com/Ruan-Xin/flink/commit/ec30671bf09b056ef47b4b89001bd8118ab23513

**Comment by Ruan-Xin** [comment]
Can run tests directly

**Comment by lirui-apache** [comment]
Thanks @Ruan-Xin for working on this. I noted there're several other types missing too, e.g. array, map. I wonder whether we can just call Hive method `ObjectInspectorFactory::getReflectionObjectInspector` to get the object inspector?

**Comment by Ruan-Xin** [comment]
> getReflectionObjectInspector

Yes, I also found out that I will take the time to make SimpleUDF support array and map types

**Comment by Ruan-Xin** [comment]
> Thanks @Ruan-Xin for working on this. I noted there're several other types missing too, e.g. array, map. I wonder whether we can just call Hive method `ObjectInspectorFactory::getReflectionObjectInspector` to get the object inspector?

I found that I can't get the type of key and value in Map type, and can't create the corresponding ObjectInspector

**Comment by lirui-apache** [comment]
@Ruan-Xin I meant you can change `HiveSimpleUDF::getHiveResultType` to something like this:
```
	@Override
	public DataType getHiveResultType(Object[] constantArguments, DataType[] argTypes) {
		try {
			List<TypeInfo> argTypeInfo = new ArrayList<>();
			for (DataType argType : argTypes) {
				argTypeInfo.add(HiveTypeUtil.toHiveTypeInfo(argType, false));
			}
			Method evalMethod = hiveFunctionWrapper.createFunction().getResolver().getEvalMethod(argTypeInfo);

			return HiveTypeUtil.toFlinkType(
					ObjectInspectorFactory.getReflectionObjectInspector(evalMethod.getGenericReturnType(),
							ObjectInspectorFactory.ObjectInspectorOptions.JAVA));
		} catch (UDFArgumentException e) {
			throw new FlinkHiveUDFException(e);
		}
	}
```
So you don't have to handle key/value types by yourself. I haven't done thorough tests for this change but seems it works for simple cases at least.

**Comment by Ruan-Xin** [comment]
> ObjectInspectorFactory.ObjectInspectorOptions.JAVA)

I got it, You're right, I can get the correct result with this method.

**Comment by Ruan-Xin** [comment]
Hi @lirui-apache  I have pushed the new fix and add test cases~

**Comment by Ruan-Xin** [comment]
Hi @lirui-apache @JingsongLi  PTAL

**Comment by lirui-apache** [comment]
@Ruan-Xin Thanks for the update. Some suggestions about the current PR:
1. Please make sure to follow flink's code style which will be verified by the CI system. You can refer to the wiki page for guidances: https://flink.apache.org/contributing/code-style-and-quality-preamble.html
2. We already have a test class for UDFs: `HiveSimpleUDFTest`. Please add new tests in that class. And you don't have to add a java file for each UDF implementation -- just use inner classes will be fine.

**Comment by Ruan-Xin** [comment]
@flinkbot run travis re-run the last Travis build
@flinkbot run azure re-run the last Azure build

**Comment by Ruan-Xin** [comment]
@flinkbot run travis re-run the last Travis build
@flinkbot run azure re-run the last Azure build

**Comment by Ruan-Xin** [comment]
Hi @JingsongLi @lirui-apache PTAL

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java [hunk 1]
```diff
@@ -354,25 +354,25 @@ public static ObjectInspector getObjectInspector(HiveShim hiveShim, Class clazz)
 		if (clazz.equals(String.class) || clazz.equals(Text.class)) {
 
 			typeInfo = TypeInfoFactory.stringTypeInfo;
-		} else if (clazz.equals(Boolean.class) || clazz.equals(BooleanWritable.class)) {
+		} else if (clazz.equals(boolean.class) || clazz.equals(Boolean.class) || clazz.equals(BooleanWritable.class)) {
```
[the remark this answers is not in the document]
Comment (Ruan-Xin) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java:357]: Ok, I will remove the method.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java [hunk 2]
```diff
@@ -49,6 +50,30 @@
...
+	@Test
+	public void testBooleanUDF() {
+		HiveSimpleUDF udf = init(BooleanUDF.class, new DataType[]{ DataTypes.INT()});
+		assertTrue((boolean) udf.eval(1));
+	}
+
+	@Test
+	public void testFloatUDF() {
+		HiveSimpleUDF udf = init(FloatUDF.class, new DataType[]{ DataTypes.FLOAT()});
+		assertEquals((float) udf.eval(3.0f), 3.0f, 1.0);
```
[the remark this answers is not in the document]
Comment (Ruan-Xin) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java:62]: the function need a positive delta

`static public void assertEquals(float expected, float actual, float delta)`

asserts that two floats are equal to within a positive delta.

I have set delta to 0

**Comment by Ruan-Xin** [comment]
@flinkbot run travis re-run the last Travis build
@flinkbot run azure re-run the last Azure build

**Comment by Ruan-Xin** [comment]
@lirui-apache PTAL

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java [hunk 3]
```diff
@@ -49,6 +50,30 @@
...
+	@Test
+	public void testBooleanUDF() {
+		HiveSimpleUDF udf = init(BooleanUDF.class, new DataType[]{ DataTypes.INT()});
+		assertTrue((boolean) udf.eval(1));
+	}
+
+	@Test
+	public void testFloatUDF() {
+		HiveSimpleUDF udf = init(FloatUDF.class, new DataType[]{ DataTypes.FLOAT()});
+		assertEquals((float) udf.eval(3.0f), 3.0f, 0);
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java:62]:
```suggestion
		assertEquals(3.0f, (float) udf.eval(3.0f), 0);
```
The 1st argument is the expected value. Same goes for other tests.

Comment (Ruan-Xin) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java:62]: ok

**Comment by Ruan-Xin** [comment]
@flinkbot run travis re-run the last Travis build
@flinkbot run azure re-run the last Azure build

**Pull request merged**